Fix background-images on table rows and sections

kohler opened the pull request:
#### d59f182d669baa4fb54eda546e9f76aa9d29ed82
<pre>
Fix background-images on table rows and sections
<a href="https://bugs.webkit.org/show_bug.cgi?id=9268">https://bugs.webkit.org/show_bug.cgi?id=9268</a>

Reviewed by Alan Baradlay

In tables, background-images applied to table rows, columns, column
groups, and row groups (sections) should span the whole relevant area.
For instance, a gradient applied to a table row should span smoothly
across the cells. This is the behavior on Firefox and Chrome, but on
Safari, the image was applied to each cell separately, resulting in
multiple distinct gradients. This commit fixes that for rows and
sections, by rendering the image with the correct dimensions, but
clipped to the relevant cells.

* LayoutTests/fast/table/row-background-image-expected.html: Added.
* LayoutTests/fast/table/row-background-image.html: Added.
* Source/WebCore/rendering/RenderTableCell.cpp:
* Source/WebCore/rendering/RenderTableCell.h:
(WebCore::RenderTableCell::paintBackgroundsBehindCell):
Add support for correct row and section background rendering.
Change paintBackgroundsBehindCell signature to take the position
relevant to the background object.
* Source/WebCore/rendering/RenderTableRow.cpp:
* Source/WebCore/rendering/RenderTableSection.cpp:
Pass new argument to paintBackgroundsBehindCell.

Canonical link: <a href="https://commits.webkit.org/282122@main">https://commits.webkit.org/282122@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/30a835137cc088d6c8b7a786e4aa3b4f141bae17

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/62150 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41504 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14742 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/66130 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12695 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/64269 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/49191 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/13035 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/50088 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8785 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/65219 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38557 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/53828 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30891 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/35232 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/11092 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11626 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56966 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/11396 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67860 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/6093 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/11165 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/57463 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/6119 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53802 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57688 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13814 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5031 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/37304 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/38388 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/39484 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/38133 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderTable.cpp:
```diff
@@ -1096,6 +1096,15 @@ LayoutUnit RenderTable::offsetHeightForColumn(const RenderTableCol& column) cons
     return m_columnOffsetHeight;
 }
 
+LayoutRect RenderTable::frameRectForColumn(const RenderTableCol& column) const
```
I think frameRect is a slightly vague name for this geometry but then again what this function returns does not really have a spec definition (size is "border box", position is certainly not). It also seems very specific to background painting, so maybe it does not even have to be on the RenderTable and instead just have a local static function or a lambda instead.

kohler replied:
The name comes from RenderBox::frameRect(). References to frameRect() are very common in the Render classes. I put the function here by analogy with offsetLeftForColumn, etc.

alanbaradlay replied:
Indeed but it is a legacy term and new code should help us converge to a  more spec friendly codebase (and also "frame position" here seems to mean (slightly) different than what  RenderBox has) -but again, this does not necessarily have to show up at RenderTable.

kohler replied:
OK. The newer version of this PR does not have this code at all.

alanbaradlay commented on Source/WebCore/rendering/RenderTableCell.cpp:
```diff
@@ -1379,17 +1379,36 @@ void RenderTableCell::paintBackgroundsBehindCell(PaintInfo& paintInfo, const Lay
 
+    // Background images attached to the row or col must span the row or col.
+    // Draw them at the row/col's dimensions but clipped to this cell.
+    bool fillBGObject = backgroundObject != this && bgLayer.hasImage();
```
please use full words (https://webkit.org/code-style-guidelines/#names-full-words)

alanbaradlay commented on Source/WebCore/rendering/RenderTableCell.cpp:
```diff
@@ -1379,17 +1379,36 @@ void RenderTableCell::paintBackgroundsBehindCell(PaintInfo& paintInfo, const Lay
+        if (auto* colObject = dynamicDowncast<RenderTableCol>(backgroundObject))
+            fillRect = colObject->computedFrameRect();
+        else if (auto* sectionObject = dynamicDowncast<RenderTableSection>(backgroundObject))
+            fillRect = LayoutRect { LayoutPoint { }, sectionObject->size() };
```
nit this could just be  `fillRect = { { }, sectionObject->size() };`

alanbaradlay commented on Source/WebCore/rendering/RenderTableCell.cpp:
```diff
@@ -1379,17 +1399,31 @@ void RenderTableCell::paintBackgroundsBehindCell(PaintInfo& paintInfo, const Lay
     }
+    LayoutRect fillRect;
+    if (paintBackgroundObject) {
+        if (auto* sectionElt = dynamicDowncast<RenderTableSection>(backgroundObject))
```
what's sectionElt?

kohler replied:
Not sure what you mean? `sectionElt` is `backgroundObject`, in the case that that element is a RenderTableSection. If you are criticizing the name, please provide an alternative.

kohler replied:
Should there be a comment added to explain the logic?

alanbaradlay replied:
tableSectionRenderer may be a better candidate (1. this is a renderer and not an element, 2. [full words please](https://webkit.org/code-style-guidelines/#names-full-words))

alanbaradlay commented on Source/WebCore/rendering/RenderTableCell.h:
```diff
@@ -83,7 +83,7 @@ class RenderTableCell final : public RenderBlockFlow {
 
     void paintCollapsedBorders(PaintInfo&, const LayoutPoint&);
-    void paintBackgroundsBehindCell(PaintInfo&, const LayoutPoint&, RenderElement* backgroundObject);
+    void paintBackgroundsBehindCell(PaintInfo&, const LayoutPoint& paintOffset, RenderBox* backgroundObject, const LayoutPoint& backgroundPaintOffset);
```
nit given the size of LayoutPoint, passing it by value is considered more performant.

kohler replied:
would you like me to change the existing LayoutPoint argument as well (it was a `const LayoutPoint&`)

alanbaradlay replied:
that would be great. Thank you!

kohler committed:
Fix background-images on table rows and sections
https://bugs.webkit.org/show_bug.cgi?id=9268

Reviewed by Alan Baradlay

In tables, background-images applied to table rows, columns, column
groups, and row groups (sections) should span the whole relevant area.
For instance, a gradient applied to a table row should span smoothly
across the cells. This is the behavior on Firefox and Chrome, but on
Safari, the image was applied to each cell separately, resulting in
multiple distinct gradients. This commit fixes that for rows and
sections, by rendering the image with the correct dimensions, but
clipped to the relevant cells.

* LayoutTests/fast/table/row-background-image-expected.html: Added.
* LayoutTests/fast/table/row-background-image.html: Added.
* Source/WebCore/rendering/RenderTableCell.cpp:
* Source/WebCore/rendering/RenderTableCell.h:
(WebCore::RenderTableCell::paintBackgroundsBehindCell):
Add support for correct row and section background rendering.
Change paintBackgroundsBehindCell signature to take the position
relevant to the background object.
* Source/WebCore/rendering/RenderTableRow.cpp:
* Source/WebCore/rendering/RenderTableSection.cpp:
Pass new argument to paintBackgroundsBehindCell.

Canonical link: https://commits.webkit.org/282122@main
